### ISSUE TITLE
Delete temporary variables from memory, but maintain them on graph for back propagation. 

### ISSUE THREAD

**Issue by myy920213** [tensorflow/tensorflow#33445]
**System information**
- TensorFlow version: 1.15

**Describe the feature and the current behavior/state.**
I'm wondering if there is any feature that allows deleting a temporary variable from memory but still keep it on the graph. For example, I'm writing a custom layer:
```bash
from tensorflow.keras import layers


class custom_layer(layers.Layer):

  def __init__(self, units=32, input_dim=32):
    super(custom_layer, self).__init__()
    self.units = units
    self.input_dim = input_dim
    w_init = tf.random_normal_initializer()
    self.w = tf.Variable(initial_value=w_init(shape=(self.input_dim, self.units),
                                              dtype='float32'),
                         trainable=True)
    b_init = tf.zeros_initializer()
    self.b = tf.Variable(initial_value=b_init(shape=(self.units,),
                                              dtype='float32'),
                         trainable=True)

  def call(self, inputs1, inputs2):
    batch_size = tf.shape(inputs1)[0]
    G = tf.matrix_diag(tf.matmul(inputs1, self.w) + self.b)
    map = tf.matrix_inverse(tf.eye(self.units, batch_shape=[batch_size]) + tf.matmul(G, tf.transpose(G, perm = [0,2,1])))
    res = tf.matmul(inputs2, map)
    
    return res

```    

The tensor ```G``` and ```map``` will take large memory space here, so I want to delete these variables to free the memory they took but still keep them on graph. When going through back propagation, if any of these variable is needed, it still can be recomputed according to the forward graph. I know this will definitely cost longer computation time but I'm willing to trade time for space. Just curious whether there is any feature like this.
 Or any suggestion about trade time for space will be appreciated.

**Who will benefit with this feature?**
People who doing research on representation learning, meta learning.

**Comment by robieta** [comment]
The API that you're looking for is `tf.custom_gradient`, and the technique is called gradient checkpointing. There's no need to use tf.Variables to manage when values are collected; the standard Tensor lifetime analysis will do that. (Since the gradients will mostly be using rematerialized values rather than the original Tensors from the forward pass.)
